### PR TITLE
feat: add boundaries_in_proof to return all boundary keys

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Summary

- Adds `boundaries_in_proof(proof_bytes)` at the merk level — returns all `KVDigest`/`KVDigestCount` keys from proof bytes
- Adds `GroveDBProof::boundaries(path)` at the grovedb level — navigates to the correct layer and delegates to merk
- Complements the existing `key_exists_as_boundary` (single key check) with a bulk variant

## Test plan

- [x] `RangeAfter` — boundary key 3 found in returned vec
- [x] `RangeAfterTo` — boundary key 2 found, result keys absent
- [x] Inclusive `Range` — no boundaries returned for result keys
- [x] Empty proof bytes — returns empty vec
- [x] GroveDB round-trip — `boundaries()` returns key 3, excludes result keys 4 and 5

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Expose a function to extract all boundary keys from proofs, allowing users to enumerate and check boundary elements that anchor exclusive-range queries.

* **Documentation**
  * Added comprehensive "Boundary Key Detection" docs (multi-language) covering when boundaries appear, how to retrieve and verify them, pagination use cases, and security caveats. Note: content appears duplicated in two spots.

* **Tests**
  * Added integration and unit tests validating boundary extraction and behavior across query types.

### CODE REVIEW

File: docs/book/translations/es/src/proof-system.md [hunk 1]
```diff
@@ -411,6 +411,85 @@ graph TD
...
+> escaneo sintactico de los nodos `KVDigest`/`KVDigestCount` de la prueba. No
+> proporcionan ninguna garantia criptografica por si solos — siempre verifique la
+> prueba contra un hash raiz de confianza primero. Los mismos tipos de nodos tambien
+> aparecen en pruebas de ausencia, por lo que el llamador debe interpretar los
+> resultados en el contexto de la consulta que genero la prueba.
+
+A nivel de merk, las mismas verificaciones estan disponibles a traves de
+`boundaries_in_proof(proof_bytes)` y
+`key_exists_as_boundary_in_proof(proof_bytes, key)` para trabajar directamente con
+los bytes de prueba merk sin procesar.
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply docs/book/translations/es/src/proof-system.md]: This is Claude. Fixed in commit fe00c662.

File: docs/book/translations/fr/src/proof-system.md [hunk 2]
```diff
@@ -411,6 +411,83 @@ graph TD
...
+> syntaxique des noeuds `KVDigest`/`KVDigestCount` de la preuve. Ils ne fournissent
+> aucune garantie cryptographique a eux seuls — verifiez toujours la preuve contre un
+> hachage racine de confiance en premier. Les memes types de noeuds apparaissent aussi dans
+> les preuves d'absence, donc l'appelant doit interpreter les resultats dans le contexte
+> de la requete qui a genere la preuve.
+
+Au niveau merk, les memes verifications sont disponibles via
+`boundaries_in_proof(proof_bytes)` et
+`key_exists_as_boundary_in_proof(proof_bytes, key)` pour travailler directement avec
+les octets bruts de la preuve merk.
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply docs/book/translations/fr/src/proof-system.md]: This is Claude. Fixed in commit fe00c662.

File: docs/book/translations/pl/src/proof-system.md [hunk 3]
```diff
@@ -415,6 +415,86 @@ graph TD
...
+sie w dowodach dla wylaczajacych typow zapytan zakresowych:
+
+| Typ zapytania | Klucz graniczny | Co dowodzi |
+|------------|-------------|----------------|
+| `RangeAfter(start..)` | `start` | Wylaczajacy poczatek istnieje w drzewie |
+| `RangeAfterTo(start..end)` | `start` | Wylaczajacy poczatek istnieje w drzewie |
+| `RangeAfterToInclusive(start..=end)` | `start` | Wylaczajacy poczatek istnieje w drzewie |
+
+Wezly graniczne pojawiaja sie rowniez w dowodach nieobecnosci, gdzie sasiednnie
+klucze dowodza istnienia luki (patrz [Dowody nieobecnosci](#dowody-nieobecnosci)
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply docs/book/translations/pl/src/proof-system.md]: This is Claude. Fixed in commit fe00c662.

File: docs/book/translations/th/src/proof-system.md [hunk 4]
```diff
@@ -399,6 +399,82 @@ graph TD
...
+ใน proof สำหรับประเภทคิวรีช่วงแบบ exclusive:
+
+| ประเภทคิวรี | Boundary key | สิ่งที่พิสูจน์ |
+|------------|-------------|----------------|
+| `RangeAfter(start..)` | `start` | จุดเริ่มต้นแบบ exclusive มีอยู่ในต้นไม้ |
+| `RangeAfterTo(start..end)` | `start` | จุดเริ่มต้นแบบ exclusive มีอยู่ในต้นไม้ |
+| `RangeAfterToInclusive(start..=end)` | `start` | จุดเริ่มต้นแบบ exclusive มีอยู่ในต้นไม้ |
+
+Boundary node ยังปรากฏใน absence proof ที่ key ข้างเคียงพิสูจน์ว่ามีช่องว่าง
+อยู่ (ดู [Absence Proof](#absence-proofs) ด้านบน)
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply docs/book/translations/th/src/proof-system.md]: This is Claude. Fixed in commit fe00c662.

File: docs/book/translations/pl/src/proof-system.md [hunk 5]
```diff
@@ -415,6 +415,86 @@ graph TD
...
+
+```rust
+let cursor_exists = grovedb_proof
+    .key_exists_as_boundary(&[b"documents", b"notes"], &cursor_key)?;
+```
+
+### Praktyczne zastosowanie: weryfikacja paginacji
+
+Jest to szczegolnie przydatne przy **paginacji**. Gdy klient zada „nastepnych
+100 dokumentow po dokumencie X", zapytanie to `RangeAfter(document_X_id)`.
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply docs/book/translations/pl/src/proof-system.md]: This is Claude. Fixed in commit ff3f28fc.

File: grovedb/src/operations/proof/mod.rs [hunk 6]
```diff
@@ -417,6 +417,59 @@ impl GroveDBProof {
...
+        if depth == path.len() {
+            return grovedb_merk::proofs::query::boundaries_in_proof(&layer.merk_proof)
+                .map_err(Into::into);
+        }
+        let segment = path[depth];
+        match layer.lower_layers.get(segment) {
+            Some(child_layer) => Self::boundaries_in_merk_layer(child_layer, path, depth + 1),
+            None => Err(Error::InvalidInput("path segment not found in proof layer")),
+        }
+    }
```
Comment (thepastaclaw) [inline grovedb/src/operations/proof/mod.rs:447]: **🟡 Suggestion: V0 proof path (boundaries_in_merk_layer) has zero test coverage**

All grovedb integration tests use `GroveVersion::latest()` which produces V1 proofs via `boundaries_in_layer`. The V0 code path through `boundaries_in_merk_layer` (lines 433-447) is never exercised by any test. This is the primary cause of any codecov shortfall on the new code. Adding a single test that constructs a V0 proof and calls `boundaries()` would cover this path.

<sub>source: ['claude-general', 'claude-rust-quality']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `grovedb/src/operations/proof/mod.rs`:
- [SUGGESTION] lines 433-447: V0 proof path (boundaries_in_merk_layer) has zero test coverage
  All grovedb integration tests use `GroveVersion::latest()` which produces V1 proofs via `boundaries_in_layer`. The V0 code path through `boundaries_in_merk_layer` (lines 433-447) is never exercised by any test. This is the primary cause of any codecov shortfall on the new code. Adding a single test that constructs a V0 proof and calls `boundaries()` would cover this path.
```
</details>

File: merk/src/proofs/query/merk_integration_tests.rs [hunk 7]
```diff
@@ -4493,3 +4493,141 @@ fn test_inclusive_range_no_boundary_at_start() {
...
+        boundaries.contains(&vec![2]),
+        "Key 2 should be boundary, got: {:?}",
+        boundaries
+    );
+    assert!(
+        boundaries.contains(&vec![5]),
+        "Key 5 should be boundary, got: {:?}",
+        boundaries
+    );
+}
```
Comment (thepastaclaw) [inline merk/src/proofs/query/merk_integration_tests.rs:4633]: **🟡 Suggestion: No test exercises KVDigestCount extraction in boundaries_in_proof**

All merk-level `boundaries_in_proof` tests use `make_6_node_tree()` which inserts plain items, producing only `KVDigest` boundary nodes. The `KVDigestCount` branch (lines 714-715 of verify.rs) — used when trees have `ProvableCountTree` feature type — is untested. A test using a count-tree or sum-tree with count would cover this.

<sub>source: ['claude-general', 'claude-rust-quality']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `merk/src/proofs/query/merk_integration_tests.rs`:
- [SUGGESTION] lines 4497-4633: No test exercises KVDigestCount extraction in boundaries_in_proof
  All merk-level `boundaries_in_proof` tests use `make_6_node_tree()` which inserts plain items, producing only `KVDigest` boundary nodes. The `KVDigestCount` branch (lines 714-715 of verify.rs) — used when trees have `ProvableCountTree` feature type — is untested. A test using a count-tree or sum-tree with count would cover this.
```
</details>

File: merk/src/proofs/query/verify.rs [hunk 8]
```diff
@@ -699,3 +699,24 @@ pub fn key_exists_as_boundary_in_proof(proof_bytes: &[u8], key: &[u8]) -> Result
...
+            | Op::PushInverted(Node::KVDigest(k, _))
+            | Op::Push(Node::KVDigestCount(k, _, _))
+            | Op::PushInverted(Node::KVDigestCount(k, _, _)) => {
+                keys.push(k);
+            }
+            _ => {}
+        }
+    }
+    Ok(keys)
+}
```
Comment (thepastaclaw) [inline merk/src/proofs/query/verify.rs:722]: **🟡 Suggestion: boundaries_in_proof duplicates decoder loop from key_exists_as_boundary_in_proof**

Lines 684-701 (`key_exists_as_boundary_in_proof`) and 706-722 (`boundaries_in_proof`) share a nearly identical decoder loop with only the accumulator differing. If either function's match arms are updated (e.g., to handle a new node type), the other must be updated in lockstep or they'll silently diverge.

Note: The suggested refactoring of `key_exists_as_boundary_in_proof` to delegate to `boundaries_in_proof` (as proposed by one agent) would introduce unnecessary allocation and lose the early-return optimization. A better approach if deduplication is desired would be a shared internal iterator/callback pattern, or simply a code comment linking the two functions.

<sub>source: ['claude-general', 'claude-rust-quality']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `merk/src/proofs/query/verify.rs`:
- [SUGGESTION] lines 706-722: boundaries_in_proof duplicates decoder loop from key_exists_as_boundary_in_proof
  Lines 684-701 (`key_exists_as_boundary_in_proof`) and 706-722 (`boundaries_in_proof`) share a nearly identical decoder loop with only the accumulator differing. If either function's match arms are updated (e.g., to handle a new node type), the other must be updated in lockstep or they'll silently diverge.

Note: The suggested refactoring of `key_exists_as_boundary_in_proof` to delegate to `boundaries_in_proof` (as proposed by one agent) would introduce unnecessary allocation and lose the early-return optimization. A better approach if deduplication is desired would be a shared internal iterator/callback pattern, or simply a code comment linking the two functions.
```
</details>

File: grovedb/src/operations/proof/mod.rs [hunk 9]
```diff
@@ -417,6 +417,59 @@ impl GroveDBProof {
...
+            };
+            return grovedb_merk::proofs::query::boundaries_in_proof(merk_bytes)
+                .map_err(Into::into);
+        }
+        let segment = path[depth];
+        match layer.lower_layers.get(segment) {
+            Some(child_layer) => Self::boundaries_in_layer(child_layer, path, depth + 1),
+            None => Err(Error::InvalidInput("path segment not found in proof layer")),
+        }
+    }
```
Comment (thepastaclaw) [inline grovedb/src/operations/proof/mod.rs:471]: **💬 Nitpick: Four near-identical recursive layer traversal helpers**

`boundaries_in_merk_layer`, `boundaries_in_layer`, `find_boundary_in_merk_layer`, and `find_boundary_in_layer` all share the same recursive descent pattern (check depth, get segment, recurse into child). This mirrors the existing V0/V1 structural split so it's understandable, but a generic helper parameterized by the leaf action could collapse all four. Low priority given the pattern is already established.

<sub>source: ['claude-rust-quality']</sub>

File: grovedb/src/operations/proof/mod.rs [hunk 10]
```diff
@@ -417,6 +417,59 @@ impl GroveDBProof {
         }
     }
 
+    /// Returns all boundary keys (`KVDigest` and `KVDigestCount` nodes)
+    /// found at the specified path in this proof.
+    ///
+    /// **Important:** This performs a syntactic scan of proof nodes. It
+    /// provides no cryptographic guarantee on its own — the proof should
+    /// be verified against a trusted root hash first.
```
Comment (thepastaclaw) [inline grovedb/src/operations/proof/mod.rs:425]: **💬 Nitpick: Doc comment could mention returned Vec may contain duplicate keys**

On a well-formed proof the same key won't appear as a boundary twice, but the doc already warns this is a syntactic scan with no cryptographic guarantee. If called on a malformed or unverified proof, duplicates are theoretically possible. Mentioning this explicitly would be a minor clarity improvement, but the existing warning is arguably sufficient.

<sub>source: ['claude-general', 'claude-rust-quality']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Clean, well-structured addition of boundaries() and boundaries_in_proof() with good test coverage at the merk level. No correctness or security bugs. The main gaps are missing test coverage for the V0 proof path and the KVDigestCount branch, which are both valid but non-blocking suggestions.

*Reviewed commit: 02e88bb4*

🟡 3 suggestion(s) | 💬 2 nitpick(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `grovedb/src/operations/proof/mod.rs`:
- [SUGGESTION] lines 433-447: V0 proof path (boundaries_in_merk_layer) has zero test coverage
  All grovedb integration tests use `GroveVersion::latest()` which produces V1 proofs via `boundaries_in_layer`. The V0 code path through `boundaries_in_merk_layer` (lines 433-447) is never exercised by any test. This is the primary cause of any codecov shortfall on the new code. Adding a single test that constructs a V0 proof and calls `boundaries()` would cover this path.

In `merk/src/proofs/query/merk_integration_tests.rs`:
- [SUGGESTION] lines 4497-4633: No test exercises KVDigestCount extraction in boundaries_in_proof
  All merk-level `boundaries_in_proof` tests use `make_6_node_tree()` which inserts plain items, producing only `KVDigest` boundary nodes. The `KVDigestCount` branch (lines 714-715 of verify.rs) — used when trees have `ProvableCountTree` feature type — is untested. A test using a count-tree or sum-tree with count would cover this.

In `merk/src/proofs/query/verify.rs`:
- [SUGGESTION] lines 706-722: boundaries_in_proof duplicates decoder loop from key_exists_as_boundary_in_proof
  Lines 684-701 (`key_exists_as_boundary_in_proof`) and 706-722 (`boundaries_in_proof`) share a nearly identical decoder loop with only the accumulator differing. If either function's match arms are updated (e.g., to handle a new node type), the other must be updated in lockstep or they'll silently diverge.

Note: The suggested refactoring of `key_exists_as_boundary_in_proof` to delegate to `boundaries_in_proof` (as proposed by one agent) would introduce unnecessary allocation and lose the early-return optimization. A better approach if deduplication is desired would be a shared internal iterator/callback pattern, or simply a code comment linking the two functions.
```
</details>
